### PR TITLE
Move PreserveBaseOverridesAttribute to special types.

### LINKED ISSUE #44208: Covariant returns when the platform lacks `PreserveBaseOverridesAttribute`
We need to decide how the compiler should behave when the underlying platform advertises support for covariant returns but lacks an accessible declaration for `System.Runtime.CompilerServices.PreserveBaseOverridesAttribute` or its default constructor. Some choices are:
1. Simply do not produce the attribute
2. Produce an error at compile-time
3. Produce a warning at compile-time
4. Generate an attribute definition in the source module.

The current behavior in the feature branch is (1).

### PR DESCRIPTION
**Pull request by gafter**
Fixes #44208

### CODE REVIEW

**Comment by gafter** [comment]
@dotnet/roslyn-compiler May I please have a couple of reviews of this?

**Comment by gafter** [comment]
@jaredpar Could you please help shepherd this PR through merge? If there is any substantial work involved and it can wait I'll deal with it later.

File: src/Compilers/Core/Portable/SpecialType.cs [hunk 1]
```diff
@@ -255,9 +255,14 @@ public enum SpecialType : sbyte
         /// </summary>
         System_Runtime_CompilerServices_RuntimeFeature = 44,
 
+        /// <summary>
+        /// An attribute that is placed on each method with a 'methodimpl" aka ".override" in metadata.
```
Comment (AlekseyTs) [inline src/Compilers/Core/Portable/SpecialType.cs:259]: >An attribute that is placed on each method with a 'methodimpl" aka ".override" in metadata. [](start = 12, length = 91)

By C# compiler? VB doesn't do that, correct?

File: src/Compilers/Core/Portable/SpecialMembers.cs [hunk 2]
```diff
@@ -1015,6 +1015,13 @@ static SpecialMembers()
                 (byte)SpecialType.System_Runtime_CompilerServices_RuntimeFeature,                                           // DeclaringTypeId
                 0,                                                                                                          // Arity
                     (byte)SignatureTypeCode.TypeHandle, (byte)SpecialType.System_String,                                    // Field Signature
+
+                // System_Runtime_CompilerServices_PreserveBaseOverridesAttribute__ctor
+                (byte)MemberFlags.Constructor,                                                                              // Flags
+                (byte)SpecialType.System_Runtime_CompilerServices_PreserveBaseOverridesAttribute,                              // DeclaringTypeId
```
Comment (AlekseyTs) [inline src/Compilers/Core/Portable/SpecialMembers.cs:1021]: >   // DeclaringTypeId [](start = 124, length = 21)

The comment looks misaligned. Inconsistent with the members above.

File: src/Compilers/VisualBasic/Test/Symbol/SymbolsTests/CompilationCreationTests.vb [hunk 3]
```diff
@@ -95,7 +95,8 @@ Namespace Microsoft.CodeAnalysis.VisualBasic.UnitTests
             For i As Integer = 1 To SpecialType.Count Step 1
                 Dim type As NamedTypeSymbol = c1.Assembly.GetSpecialType(CType(i, SpecialType))
 
-                If i = SpecialType.System_Runtime_CompilerServices_RuntimeFeature Then
+                If i = SpecialType.System_Runtime_CompilerServices_RuntimeFeature Or
+                    i = SpecialType.System_Runtime_CompilerServices_PreserveBaseOverridesAttribute Then
```
Comment (AlekseyTs) [inline src/Compilers/VisualBasic/Test/Symbol/SymbolsTests/CompilationCreationTests.vb:99]: >i [](start = 20, length = 1)

Consider aligning with `i` above

File: src/Compilers/VisualBasic/Test/Symbol/SymbolsTests/WellKnownTypeValidationTests.vb [hunk 4]
```diff
@@ -454,7 +454,8 @@ End Namespace
                 Dim symbol = comp.GetSpecialType(special)
                 Assert.NotNull(symbol)
 
-                If special = SpecialType.System_Runtime_CompilerServices_RuntimeFeature Then
+                If special = SpecialType.System_Runtime_CompilerServices_RuntimeFeature OrElse
+                    special = SpecialType.System_Runtime_CompilerServices_PreserveBaseOverridesAttribute Then
```
Comment (AlekseyTs) [inline src/Compilers/VisualBasic/Test/Symbol/SymbolsTests/WellKnownTypeValidationTests.vb:458]: >special [](start = 20, length = 7)

Consider aligning with `special` above

File: src/Compilers/Core/Portable/SpecialType.cs (hunk 1 above)
Comment (gafter) [reply src/Compilers/Core/Portable/SpecialType.cs:259]: VB doesn't current do that, but it could and should.

**Comment by gafter** [comment]
@jcouv: regarding
> I didn't see PreserveBaseOverridesAttribute mentioned in https://github.com/dotnet/csharplang/blob/master/proposals/csharp-9.0/covariant-returns.md
> Can we update the spec?

That link is a language specification. The attribute is not part of the C# language changes. That attribute is part of the CLR support and is specified in draft CLR changes.

**Comment by jcouv** [comment]
Thanks, found it: https://github.com/dotnet/runtime/blob/master/docs/design/features/covariant-return-methods.md
Added that link to our test plan for reference

**Pull request merged**